Add driver is31fl3194

benjaminbjornsson opened the pull request.

benjaminbjornsson committed:
boards: arm: arduino_nicla_sense_me: Add IS31FL3194 LED driver to DTS

This commit adds the IS31FL3194 LED driver to the board DTS.

Signed-off-by: Benjamin Björnsson <benjamin.bjornsson@gmail.com>

benjaminbjornsson committed:
samples: drivers: led: Add sample for IS31FL3194 LED driver

This commit add a sample to demonstrate basic usage
of the IS31FL3194 LED driver.

Signed-off-by: Benjamin Björnsson <benjamin.bjornsson@gmail.com>

carlescufi commented:
@mbolivar-nordic @MaureenHelm could you take a look?

simonguinot commented on dts/bindings/led/issi,is31fl3194.yaml:
```diff
@@ -0,0 +1,32 @@
+include: ["i2c-device.yaml", "led-controller.yaml"]
+
+child-binding:
+    description: TI LP503x LED controller node
```
I don't think it is the right description :)

benjaminbjornsson replied:
Yeah, that doen't look right... Thank you!

simonguinot commented:
Hi @carlescufi. I'll review it in the next few days.

benjaminbjornsson commented:
> Hi @carlescufi. I'll review it in the next few days.

Thank you @simonguinot !

mbolivar-nordic commented:
> @mbolivar-nordic [...] could you take a look?

I will try, but this is led, not led_strip, so @Mani-Sadhasivam is the maintainer

Mani-Sadhasivam commented:
It's on the pipeline for review this weekend.

simonguinot commented on drivers/led/is31fl3194/is31fl3194.c:
```diff
@@ -0,0 +1,235 @@
+#define IS31FL3194_INDEX_MAX 2
+
+struct is31fl3194_config {
+	struct is31fl3194_ctx ctx;
```
I am not sure, but I think this could be `const` because the pointers inside are not meant to be written. Isn't it ?

benjaminbjornsson replied:
I don't think const is necessary inside `struct is31fl3194_config`. The entire struct will be declared as `const` when used in the device macro.

simonguinot replied:
OK

simonguinot commented on drivers/led/is31fl3194/is31fl3194.c:
```diff
@@ -0,0 +1,235 @@
+	.off = is31fl3194_off,
+	/* Optional callbacks. */
+	.get_info = is31fl3194_get_info,
+	.set_brightness = is31fl3194_set_brightness,
```
Should you not implement the set_color function ? It looks like configuring the color is supported by the IS31FL3194 LED driver.

benjaminbjornsson replied:
I think the reason why I didn't is that `set_color` sets the color of a certain LED whereas this driver has three single color LEDs connected to it.

simonguinot replied:
Are there 3 LEDs ? Or a single LED with 3 channels ?

In the former case, a user may need to configure the LED by calling a single function (with a single I2C command) rather than setting the brightness of three different LED which are in fact a single physical one. This makes a difference when running software LED effects.

Note that the write_channels function can also be used to do that, but in a much less intuitive way.

simonguinot replied:
According to the datasheet the is31fl3194 is a 3-channel LED driver and both options are possible: 3 single color LEDs, or a RGB / RG+W LED. So maybe you should implement `set_color` as well. It is your call and you can consider this comment resolved either way.

simonguinot commented on drivers/led/is31fl3194/is31fl3194.c:
```diff
@@ -0,0 +1,235 @@
+	return 0;
+}
+
+int is31fl3194_set_brightness(const struct device *dev, uint32_t led, uint8_t value)
```
In led.h it is written that the brightness value is a percent. So I think you have to convert the brightness value from the [0-100] range to [0-255].

Note that the API should be changed because there is no reason to loose granularity on brightness.

benjaminbjornsson replied:
Agreed, will updated this.

I agree, makes more sense to use full range of `uint8_t`.

simonguinot commented on drivers/led/is31fl3194/is31fl3194.c:
```diff
@@ -0,0 +1,235 @@
+		.color_mapping = color_mapping_##led_node_id,	\
+	},
+
+#define IS31FL3194_CURR_BAND_MAX(id) DT_PROP_OR(id, curr_map_max, 20),
```
I don't think it is working. The property is named `curr-band-max` in the DT binding. So it should be `curr_band_max` here.

simonguinot commented on dts/bindings/led/issi,is31fl3194.yaml:
```diff
@@ -0,0 +1,32 @@
+            - 2
+        color-mapping:
+          required: true
+        curr-band-max:
```
Please, can you add some description for this property ?

simonguinot commented on drivers/led/is31fl3194/is31fl3194_reg.h:
```diff
@@ -0,0 +1,483 @@
+#define IS31FL3194_COLOR_3_P2_BASE 0x26U
+#define IS31FL3194_COLOR_3_P3_BASE 0x36U
+struct is31fl3194_px_coly_oz {
+	uint8_t pxcoly_oz : 8;
```
I think you can remove the `: 8` here :)

benjaminbjornsson replied:
I think it's quite nice to state explicitly that we're using the whole `uint8_t` here and IMO it makes sense to be consistent across structs.

simonguinot replied:
OK

simonguinot commented on drivers/led/is31fl3194/is31fl3194_reg.c:
```diff
@@ -0,0 +1,837 @@
+
+static int read_reg(const struct is31fl3194_ctx *ctx, uint8_t reg, uint8_t *data, uint16_t len)
+{
+	return ctx->read_reg(ctx->handle, reg, data, len);
```
As I said in another comment, I don't see the purpose for this indirection.

Mani-Sadhasivam replied:
NACK. These kind of HAL abstractions makes sense if you are creating a library that might be used by different drivers. But here, there is obviously one driver making use of it and the HAL indirection is really making the code look complex. Please get rid of it.

simonguinot reviewed (dismissed):
Hi @benjaminbjornsson,

You really did a nice job with this driver. The IS31FL3194 is a beast. I can't wait to see how you will implement the hardware LED effects (such as breathing). Probably you already have some ideas how we could add support for LED effects to the LED API.

Note that I didn't review completely this PR. It was too big for the time I had. If needed I'll do another pass.

See below for my comments.

benjaminbjornsson commented on drivers/led/is31fl3194/is31fl3194.c:
```diff
@@ -0,0 +1,235 @@
+	struct is31fl3194_ctx *ctx = (struct is31fl3194_ctx *)&config->ctx;
+	int res;
+
+	res = is31fl3194_out_en_set(ctx, led, 1);
```
Agree, it currently just enables it. Does turning it on mean that it should be set to full brightness?

simonguinot replied:
Since it is not specified, it is up to you. But I think that full brightness should be right. Note that I am not a big fan of the `led_off` / `led_on` functions. They don't exist in the Linux LED API for example. I think of them pretty much as wrappers to `led_set_brightness` with brightness set to 0 or 100.

benjaminbjornsson commented on drivers/led/is31fl3194/is31fl3194.c:
```diff
@@ -0,0 +1,235 @@
+int is31fl3194_on(const struct device *dev, uint32_t led)
+{
+	const struct is31fl3194_config *config = dev->config;
+	struct is31fl3194_ctx *ctx = (struct is31fl3194_ctx *)&config->ctx;
```
Agreed, will add const here.

simonguinot commented:
@benjaminbjornsson I think it would be nice to provides a DTS overlay example along with the sample (maybe for the board you are using to test the driver).

benjaminbjornsson committed:
drivers: led: is31fl3194: Add support for IS31FL3194 LED Driver

This commit adds support for the IS31FL3194 LED driver.
The driver can be found on the Arduino Nicla Sense ME board
which has recently been supported.

The driver functionallity has been verified on the Arduino
Nicla Sense ME.

Signed-off-by: Benjamin Björnsson <benjamin.bjornsson@gmail.com>

benjaminbjornsson commented:
I've updated the PR and I think I've fixed all issues raised by @simonguinot. Decided to remove the auto-breathing functionality from the abstraction layer to make the PR easier to review.

benjaminbjornsson commented:
> @benjaminbjornsson I think it would be nice to provides a DTS overlay example along with the sample (maybe for the board you are using to test the driver).

Isn't the node provided in the Arduino Nicla Sense ME board DTS that is part of this PR enough as an example? Or should that node be moved into an overlay of the example?

simonguinot commented:
> > @benjaminbjornsson I think it would be nice to provides a DTS overlay example along with the sample (maybe for the board you are using to test the driver).
> 
> Isn't the node provided in the Arduino Nicla Sense ME board DTS that is part of this PR enough as an example? Or should that node be moved into an overlay of the example?

Yes it is enough.

simonguinot commented:
Sorry, I closed this PR by accident.

simonguinot reviewed (commented):
@benjaminbjornsson I still don't like the HAL idea and especially the indirection around the I2C functions (i.e. `ctx->{read,write}_reg()` -> `is31fl3194_i2c_{read,write}()` -> `i2c_burst_{read,write}_dt()`). IMHO it only makes sense if you plan to use your HAL with another OS. But I can't see the gain for Zephyr. This complicates the code. And my advice is still that you should simplify the driver.

I'll be in vacations for the next three weeks and I don't want to block this PR. So I am withdrawing my initial "request changes" feedback. But I am not approving it either.

simonguinot commented:
Hi @Mani-Sadhasivam. Please can you have a look ? What do you think of the "I2C abstraction" code in `is31fl3194_reg.{c,h}` ?

Mani-Sadhasivam commented:
> Hi @Mani-Sadhasivam. Please can you have a look ? What do you think of the "I2C abstraction" code in `is31fl3194_reg.{c,h}` ?

Thanks, @simonguinot for the reviews, really appreciate your help :)  As I said in the review, I'm not in favour of adding the HAL abstraction to this driver. So that part needs to be reworked.
